AEP/aep-00004.md

```diff
@@ -0,0 +1,187 @@
+
+No additional API is required, and few adjustments are required infrastructure-wise.
+The ingress in front of the actual cluster should accept authentication headers from the Load Balancer itself, or the Load Balancer is an authenticated user of the cluster that is able to impersonate any user of the cluster.
+
+The Load Balancer could scale to many servers without issues.
+
+## Implementation
+
+The implementation of such a Load Balancer would consist in a gRPC server and client.
+The server would take ArmoniK requests from the user, and forward them to the right cluster using ArmoniK API.
+It would forward the response back to the user.
+
+In particular, it is transparent from both ends, client and control plane.
+No specific API is required.
+
+The choice of a cluster can be random, or follow some business rules like selecting the cluster with the less load, the cheapest one, or the cluster with the fastest hardware.
+It could also be possible to select a specific cluster when creating the session using the default `TaskOptions` of the session and a known field if required.
+
+### Caching and cluster discovery
+
+When a new session is created, the link between the session ID and the cluster is recorded either in a database or in a cache (possibly within the process itself).
+This enable the Load Balancer to directly know to which cluster to redirect future requests for this session.
+There is no consideration of cache coherency between Load Balancers or cache expiration as the binding to a cluster cannot change ever.
+
+For the RPCs that do not have the session ID in their requests (like `GetTaskRequest`), a discovery among the clusters is required.
+Upon such request, the Load Balancer would forward the request to all clusters, and return only the response of the cluster containing the requested task or result.
+The link between the task/result and the cluster can also be cached to reduce the number of discoveries.
+
+To reduce even further the number of discoveries, the link between tasks/results and clusters can also be cached upon task/result creation when the cluster actually returns the new ID to the Load Balancer.
+And by adding the session ID to all the RPC, we would be able to remove all discoveries during normal execution, even when workers create new tasks and results.
+
+This principle of cluster discovery can also be used to handle Load Balancer restarts, and make them stateless.
+On startup, the cache will be empty, and all requests would trigger a discovery, even if the session ID is included in the request.
+Using a database instead of (or in addition to) a cache would avoid too many discoveries on startups, but could limit the scaling to what the database can handle.
+
+It would be possible to enable communications between Load-Balancers in order to reduce the number of discoveries altogether, at the price of a much more complicated Load Balancer.
```

Comment:
```suggestion
It could be possible to enable communications between Load-Balancers in order to reduce the number of discoveries altogether, at the price of a much more complex Load Balancer.
```